### PR TITLE
Ensure subscription for the prometheus operator was deleted

### PR DESCRIPTION
**Pull request by maknop**
ENO checks to ensure that the Subscription was deleted by Clowder, if the subscription was not, it will requeue. Once the ENO does not see the subscription anymore, the prometheus operator is deleted.

### CODE REVIEW

File: controllers/cloud.redhat.com/namespacereservation_controller.go [hunk 1]
```diff
@@ -108,6 +108,16 @@ func (r *NamespaceReservationReconciler) Reconcile(ctx context.Context, req ctrl
 			return ctrl.Result{}, err
 		}
 		if r.Poller.namespaceIsExpired(expirationTS) {
+			err = DeleteSubscriptionPrometheusOperator(ctx, r.Client, res.Status.Namespace)
+			if k8serr.IsNotFound(err) {
+				r.Log.Error(err, fmt.Sprintf("cannot find prometheus operator for namespace %s.", res.Status.Namespace))
```
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/namespacereservation_controller.go:113]: Maybe add `subscription` in here somewhere?

File: controllers/cloud.redhat.com/namespaces.go [hunk 2]
```diff
@@ -292,3 +292,38 @@ func DeletePrometheusOperator(ctx context.Context, cl client.Client, nsName stri
...
+	gvk := schema.GroupVersionKind{
+		Group:   "operators.coreos.com",
+		Version: "v1alpha1",
+		Kind:    "Subscription",
+	}
+
+	subscriptionsPrometheusOperator.SetGroupVersionKind(gvk)
+	subscriptionsPrometheusOperator.SetName(GetPrometheusOperatorName(nsName))
+
+	for {
```
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/namespaces.go:315]: Does this need to be in a loop? I might be missing some context here.

File: controllers/cloud.redhat.com/namespacereservation_controller.go [hunk 3]
```diff
@@ -108,6 +108,16 @@ func (r *NamespaceReservationReconciler) Reconcile(ctx context.Context, req ctrl
 			return ctrl.Result{}, err
 		}
 		if r.Poller.namespaceIsExpired(expirationTS) {
+			err = DeleteSubscriptionPrometheusOperator(ctx, r.Client, res.Status.Namespace)
+			if k8serr.IsNotFound(err) {
```
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/namespacereservation_controller.go:112]: It looks like you're already doing this in the `DeleteSubscriptionPrometheusOperator` method. Do we want to do that again here?

File: controllers/cloud.redhat.com/namespaces.go (hunk 2 above)
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/namespaces.go:315]: Does this need to be in a loop? I might be missing some context here.

File: controllers/cloud.redhat.com/pool_controller.go [hunk 4]
```diff
@@ -136,6 +136,16 @@ func (r *NamespacePoolReconciler) handleErrorNamespaces(ctx context.Context, err
 			return fmt.Errorf("handleErrorNamespace error: Couldn't delete namespace: %v", err)
 		}
 
+		err = DeleteSubscriptionPrometheusOperator(ctx, r.Client, nsName)
```
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/pool_controller.go:139]: We have this (mostly) same block three times. I wonder if we could extract this out to another function that's the single source.

**Comment by bsquizz** [comment]
I'm taking a look at this but am confused about something, maybe I am missing it. I see the log messages and checks to see if the subscription still exists. But at what point does the subscription get deleted? Like where is the `client.Delete` call for it? :) 

Also, when we hit the error that the subscription still exists in these various controllers, should we be returning and doing a re-queue? It seems like right now the code just prints the error message and continues -- is that the desired behavior?

**Comment by maknop** [comment]
> I'm taking a look at this but am confused about something, maybe I am missing it. I see the log messages and checks to see if the subscription still exists. But at what point does the subscription get deleted? Like where is the `client.Delete` call for it? :)
Yeah so we actually determined that Clowder will be the one to delete the subscription and that the ENO will just need to check to ensure that Clowder did deleted before proceeding to delete the prometheus operator. 

> Also, when we hit the error that the subscription still exists in these various controllers, should we be returning and doing a re-queue? It seems like right now the code just prints the error message and continues -- is that the desired behavior?
Just added the `return err`. Thanks!

**Review by psav — CHANGES_REQUESTED** [review]

File: controllers/cloud.redhat.com/namespacereservation_controller.go [hunk 5]
```diff
@@ -106,14 +106,19 @@ func (r *NamespaceReservationReconciler) Reconcile(ctx context.Context, req ctrl
...
-				r.Log.Error(err, fmt.Sprintf("Error deleting prometheus.%s", res.Status.Namespace))
-			} else {
-				r.Log.Info("Successfully deleted", "prometheus-operator", fmt.Sprintf("prometheus.%s", res.Status.Namespace))
+			} else if removed {
+				r.Log.Info("Subscription for prometheus operator was deleted", "prometheus-operator subscription", fmt.Sprint(res.Status.Namespace))
+			}
+
+			removed, err = DeletePrometheusOperator(ctx, r.Client, res.Status.Namespace)
+			if err != nil {
+				r.Log.Error(err, fmt.Sprintf("prometheus operator subscription for namespace %s still exists.", res.Status.Namespace))
```
Comment (bsquizz) [inline controllers/cloud.redhat.com/namespacereservation_controller.go:119]: Should we requeue here at this point if deleting the prometheus operator returned an err?

File: controllers/cloud.redhat.com/namespaces.go [hunk 6]
```diff
@@ -285,10 +280,40 @@ func DeletePrometheusOperator(ctx context.Context, cl client.Client, nsName stri
...
+	subscriptionsPrometheusOperator.SetNamespace(nsName)
+
+	err := cl.Get(ctx, types.NamespacedName{Name: "prometheus", Namespace: nsName}, &subscriptionsPrometheusOperator)
+	if k8serr.IsNotFound(err) {
+		return true, nil
+	} else if err == nil {
+		return false, err
+	}
+
+	return true, nil
```
Comment (bsquizz) [inline controllers/cloud.redhat.com/namespaces.go:318]: Sorry one more thing I spotted -- if we hit an err (that is not a `NotFound` error) -- we will return `true, nil`

I think we should return `false, err` in that case, right? Logic could be simplified to:

```go
	if k8serr.IsNotFound(err) {
		return true, nil
	}
        return false, err
}
```

File: controllers/cloud.redhat.com/namespaces.go [hunk 7]
```diff
@@ -285,10 +280,40 @@ func DeletePrometheusOperator(ctx context.Context, cl client.Client, nsName stri
 	prometheusOperator.SetGroupVersionKind(gvk)
 	prometheusOperator.SetName(GetPrometheusOperatorName(nsName))
 
+	err := cl.Get(ctx, types.NamespacedName{Name: GetPrometheusOperatorName(nsName)}, &prometheusOperator)
+	if k8serr.IsNotFound(err) {
+		return true, nil
+	} else if err != nil {
```
Comment (bsquizz) [inline controllers/cloud.redhat.com/namespaces.go:286]: Same kind of thing here. What if `err` is not nil but is also not a `NotFound` err? I think this should just be a plain `else {` statement.

**Review by bsquizz — CHANGES_REQUESTED** [review]

File: controllers/cloud.redhat.com/namespaces.go (hunk 7 above)
Comment (psav) [reply controllers/cloud.redhat.com/namespaces.go:286]: If you had it as a bare else, then a `nil` value for err would trigger it

**Pull request merged**